main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.unification
+
+import scala.annotation.nowarn
+import scala.collection.immutable.SortedSet
+
+object FastSetUnification {
+
+  /**
+    * A common super-type for set formulas `f`, like `x1 ∩ x2 ∪ (e4 ∪ !c17)`.
+    *
+    * A set formula can contain elements, constants (fixed unknown sets), variables (unknown sets),
+    * complements, unions, and intersections.
+    *
+    * Variables and constants are collectively referred to as unknowns.
+    * A formula without unknowns is called ground.
+    *
+    * A set without unknowns can be evaluated to get a finite or a co-finite set of elements. The
+    * universe of elements is infinite so no finite set is equivalent to universe.
+    *
+    * Invariant: [[SetFormula.ElemSet]], [[SetFormula.Cst]], and [[SetFormula.Var]] must use
+    * disjoint integers (see [[unknowns]]).
+    */
+  sealed trait SetFormula {
+
+    import SetFormula.*
+
+    /** The set of [[Var]] in `this`. */
+    final def variables: SortedSet[Int] = this match {
+      case Univ => SortedSet.empty
+      case Empty => SortedSet.empty
+      case Cst(_) => SortedSet.empty
+      case Var(x) => SortedSet(x)
+      case ElemSet(_) => SortedSet.empty
+      case Compl(t) => t.variables
+      case Inter(_, _, varsPos, _, _, varsNeg, other) =>
+        SortedSet.from(
+          varsPos.iterator.map(_.x) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.variables)
+        )
+      case Union(_, _, varsPos, _, _, varsNeg, other) =>
+        SortedSet.from(
+          varsPos.iterator.map(_.x) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.variables)
+        )
+    }
+
+    /** `true` if `this` contains neither [[Var]] nor [[Cst]]. */
+    final def ground: Boolean = this match {
+      case Univ => true
+      case Empty => true
+      case Cst(_) => false
+      case Var(_) => false
+      case ElemSet(_) => true
+      case Compl(t) => t.ground
+      case Inter(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        cstsPos.isEmpty &&
+          varsPos.isEmpty &&
+          cstsNeg.isEmpty &&
+          varsNeg.isEmpty &&
+          other.forall(_.ground)
+      case Union(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        cstsPos.isEmpty &&
+          varsPos.isEmpty &&
+          cstsNeg.isEmpty &&
+          varsNeg.isEmpty &&
+          other.forall(_.ground)
+    }
+
+    /** Returns all [[Var]] and [[Cst]] that occur in `this`. */
```

Comment:
Add a comment that the domains ofVar and Cst are assumed to be disjoint, so it is okay to return them in one set instead of in a pair of sets.

Reply:
Nevertheless, when do you need this function? Why is it not "better" to have varsOf, cstsOf etc?

Reply:
constants are never needed by themselves, so having unknowns seems better (one traversal for the union)

Its a general pattern i use to ask questions directly (t.freeVarsContains(x)) instead of general methods (t.variables.contains(x)) to have better performance